Core.UnitTests/TypeAssembly/Implementation/TypeAssemblerTest.cs

```diff
@@ -250,150 +275,155 @@ public void AssembleType_RequestAssembledType ()
+
+                additionalTypeAssemblyContext.GenerationCompleted += ctx =>
                 {
-                    new KeyValuePair<MutableType, Type> (additionalMutableType, fakeAdditionalType),
-                    new KeyValuePair<MutableType, Type> (otherAdditionalType, otherFakeAdditionalType)
-                })
-            .WhenCalled (mi => Assert.That (generationCompletedEventRaised, Is.False));
-      }
-      mockRepository.ReplayAll();
+                  Assert.That (ctx.GetGeneratedType (additionalMutableType), Is.SameAs (fakeAdditionalType));
+                  Assert.That (ctx.GetGeneratedType (otherAdditionalType), Is.SameAs (otherFakeAdditionalType));
+                  generationCompletedEventRaised = true;
+                };
+              });
+      mutableTypeFactoryMock
+          .InSequence (sequence)
+          .Setup (mock => mock.CreateType ("AdditionalType", null, 0, typeof (int), null)).Returns (additionalMutableType);
+      mutableTypeFactoryMock
+          .InSequence (sequence)
+          .Setup (mock => mock.CreateType ("OtherAdditionalType", null, 0, typeof (int), null)).Returns (otherAdditionalType);
+
+      participantMock2
+          .InSequence (sequence)
+          .Setup (mock => mock.GetOrCreateAdditionalType (additionalTypeID, It.IsAny<IAdditionalTypeAssemblyContext>()))
+          .Returns (additionalMutableType)
+          .Callback (
+              (object _, IAdditionalTypeAssemblyContext additionalTypeAssemblyContextArg) =>
+              {
+                Assert.That (additionalTypeAssemblyContextArg, Is.SameAs (additionalTypeAssemblyContextPassedToParticipant1));
+              });
+      // Participant 3 is not invoked.
+
+      codeGeneratorMock
+          .InSequence (sequence)
+          .Setup (mock => mock.GenerateTypes (new[] { additionalMutableType, otherAdditionalType }))
+          .Returns (new[] { new KeyValuePair<MutableType, Type> (additionalMutableType, fakeAdditionalType), new KeyValuePair<MutableType, Type> (otherAdditionalType, otherFakeAdditionalType) })
+          .Callback ((IEnumerable<MutableType> mutableTypes) => Assert.That (generationCompletedEventRaised, Is.False));
       var typeAssembler = CreateTypeAssembler (
-          mutableTypeFactoryMock,
+          mutableTypeFactoryMock.Object,
           configurationId: participantConfigurationID,
-          participants: new[] { participantMock1, participantMock2, participantMock3 });
+          participants: new[] { participantMock1.Object, participantMock2.Object, participantMock3.Object });
 
-      var result = typeAssembler.AssembleAdditionalType (additionalTypeID, _participantStateMock, codeGeneratorMock);
+      var result = typeAssembler.AssembleAdditionalType (additionalTypeID, _participantStateMock.Object, codeGeneratorMock.Object);
 
-      mockRepository.VerifyAll();
+      participantMock1.Verify();
+      mutableTypeFactoryMock.Verify();
+      participantMock2.Verify();
```

Comment:
verify in order of mock creation to keep symetric / easy to review. also verify pariticpantmock3